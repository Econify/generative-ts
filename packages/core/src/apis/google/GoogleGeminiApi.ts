/* eslint-disable no-nested-ternary */
/* eslint-disable camelcase */

import type { ModelApi } from "@typeDefs";

import { FnTemplate } from "../../utils/Template";

import { mapToolDescriptionsToGeminiRequest } from "./mapToolDescriptionsToGeminiRequest";
import { mapToolResultsToGeminiRequest } from "./mapToolResultsToGeminiRequest";
import {
  GoogleGeminiResponse,
  isGoogleGeminiResponse,
} from "./GoogleGeminiResponse";
import {
  GoogleGeminiOptions,
  PartWithFunctionCall,
} from "./GoogleGeminiRequest";

import { FUNCTION_CALL_WITHOUT_TOOLS } from "./errors";

/**
 * @category Google Gemini
 * @category Templates
 */
export const GoogleGeminiTemplate = new FnTemplate(
  ({
    $prompt,
    examplePairs,
    system,
    $tools,
    contents,
    tools,
    tools_config,
    system_instruction,
    safety_settings,
    generation_config,
  }: GoogleGeminiOptions) => {
    let _contents = (
      contents && Array.isArray(contents) ? contents : [contents]
    ).filter((i) => i);

    const firstItem = _contents[0];
    const lastItem = _contents[_contents.length - 1];

    // the conversation must start with a user message:
    if (!firstItem || firstItem.role === "model") {
      _contents = [
        {
          role: "user",
          parts: [{ text: $prompt }],
        },
        ..._contents,
      ];
    }

    // the conversation must end with a user message (either $prompt or tool responses):
    if (lastItem && lastItem.role === "model") {
      const functionCalls = lastItem.parts.filter(
        (part): part is PartWithFunctionCall => "functionCall" in part,
      );

      if (functionCalls.length && $tools) {
        const responses = mapToolResultsToGeminiRequest(functionCalls, {
          $tools,
        });

        if (responses.length) {
          _contents = [
            ..._contents,
            {
              role: "user",
              parts: responses,
            },
          ];
        } else {
          // if no tool responses, we logged a warning in `applyFunctionCalls`, and now append $prompt as fallback:
          _contents = [
            ..._contents,
            {
              role: "user",
              parts: [{ text: $prompt }],
            },
          ];
        }
      } else {
        if (functionCalls.length && !$tools) {
          console.warn(FUNCTION_CALL_WITHOUT_TOOLS);
        }

        _contents = [
          ..._contents,
          {
            role: "user",
            parts: [{ text: $prompt }],
          },
        ];
      }
    }

    const rewritten = {
      contents: [
        ...(examplePairs
          ? examplePairs.flatMap((pair) => [
              {
                role: "user" as const,
                parts: [{ text: pair.user }],
              },
              {
                role: "model" as const,
                parts: [{ text: pair.assistant }],
              },
            ])
          : []),
        ..._contents,
      ],
      ...(tools || $tools
        ? {
            tools: [
              ...(tools || []),
              ...($tools
                ? mapToolDescriptionsToGeminiRequest({ $tools }).tools || []
                : []),
            ],
          }
        : {}),
      ...(system_instruction || system
        ? {
            system_instruction: {
              parts: [
                ...(system ? [{ text: system }] : []),
                ...(system_instruction
                  ? system_instruction.parts.map((part) => ({
                      ...(part.text ? { text: part.text } : {}),
                    }))
                  : []),
              ],
            },
          }
        : {}),
      ...(tools_config
        ? {
            tools_config,
          }
        : {}),
      ...(safety_settings
        ? {
            safety_settings,
          }
        : {}),
      ...(generation_config
        ? {
            generation_config,
          }
        : {}),
    };

    return JSON.stringify(rewritten, null, 2);
  },
);

export interface GoogleGeminiApi
  extends ModelApi<GoogleGeminiOptions, GoogleGeminiResponse> {}

/**
 *
 * ## Reference
 * - {@link https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/inference | Gemini Inference API}
 * - {@link https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/function-calling | Gemini Function Calling API}
 *
 * ## Providers using this API
 * - {@link createVertexAiModelProvider | GCloud VertexAI}
 *
 * @category APIs
 * @category Google Gemini
 * @category Provider: GCloud VertexAI
 *
 */
export const GoogleGeminiApi: GoogleGeminiApi = {
  requestTemplate: GoogleGeminiTemplate,
  responseGuard: isGoogleGeminiResponse,
};
