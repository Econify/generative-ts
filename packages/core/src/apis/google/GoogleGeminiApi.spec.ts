import { GoogleGeminiApi } from "./GoogleGeminiApi";
import type { GoogleGeminiOptions } from "./GoogleGeminiRequest";

function render(context: Omit<GoogleGeminiOptions, "modelId">) {
  try {
    const rendered = GoogleGeminiApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("GoogleGeminiApi.requestTemplate", () => {
  /**
   * FewShotRequestOptions (prompt, examplePairs, system):
   */

  test("prompt", () => {
    const rendered = render({
      prompt: "mock-prompt",
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, system", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs, system", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system: "mock-system-text",
    });
    expect(rendered).toMatchSnapshot();
  });

  /**
   * "Native" few shot options (prompt, contents, system_instruction):
   */

  test("prompt, contents with user / model (appends prompt)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, contents with model / user (prepends prompt)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-text-2" }],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, contents starting and ending with model (appends AND prepends prompt)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-model-text-2" }],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, contents starting and ending with user (does not add prompt)", () => {
    const rendered = render({
      prompt: "mock-prompt-should-not-appear",
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-text-2" }],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, system_instruction", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, contents, system_instruction", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
      ],
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  /**
   * Combinations of FewShotRequestOptions and "native" options:
   */

  test("prompt, examplePairs, contents with user / model (appends prompt)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        {
          user: "mock-user-example-pair",
          assistant: "mock-assistant-example-pair",
        },
      ],
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs, contents with model / user (inserts prompt, conversation is valid)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        {
          user: "mock-user-example-pair",
          assistant: "mock-assistant-example-pair",
        },
      ],
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs, system_instruction", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, system, contents", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, system, system_instruction", () => {
    const rendered = render({
      prompt: "mock-prompt",
      system: "mock-system-text",
      system_instruction: {
        parts: [
          { text: "mock-additional-instruction" },
          { text: "mock-additional-instruction-2" },
        ],
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, examplePairs, contents, system_instruction", () => {
    const rendered = render({
      prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
      ],
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  /**
   * Tool-related:
   */

  test("prompt, contents ending with function_call, $tools with matching invocation (appends function_response content items)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [
            {
              functionCall: {
                name: "mock-function",
                args: { key: "value" },
              },
            },
          ],
        },
      ],
      $tools: [
        {
          name: "mock-function",
          description: "mock-description",
          parameters: [
            {
              name: "key",
              description: "mock-key-description",
              type: "STR",
              required: true,
            },
          ],
          invocations: [
            {
              arguments: { key: "value" },
              resolved: true,
              returned: { responseKey: "responseValue" },
            },
          ],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, contents ending with model function_call, $tools without matching invocation (appends prompt; TODO logs warning)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "user",
          parts: [{ text: "mock-user-text" }],
        },
        {
          role: "model",
          parts: [
            {
              functionCall: {
                name: "mock-function",
                args: { key: "value" },
              },
            },
          ],
        },
      ],
      $tools: [
        {
          name: "another-function",
          description: "another-description",
          parameters: [
            {
              name: "another-key",
              description: "another-key-description",
              type: "STR",
              required: false,
            },
          ],
          invocations: [],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
    // TODO expect warning
  });

  test("prompt, contents ending with model function_call, no $tools (prepends and appends prompt; TODO logs warning)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [
            {
              functionCall: {
                name: "mock-function",
                args: { key: "value" },
              },
            },
          ],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
    // TODO expect warning
  });

  test("prompt, contents ending with user function_response (prepends prompt)", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [
            {
              functionCall: {
                name: "mock-function",
                args: { key: "value" },
              },
            },
          ],
        },
        {
          role: "user",
          parts: [
            {
              function_response: {
                name: "mock-function",
                response: { responseKey: "responseValue" },
              },
            },
          ],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });
  /*
   *  Tool declarations:
   */
  test("prompt, tools", () => {
    const rendered = render({
      prompt: "mock-prompt",
      tools: [
        {
          function_declarations: [
            {
              name: "mock-function",
              description: "mock-description",
              parameters: {
                type: "OBJECT",
                properties: {
                  key: { type: "STRING" },
                },
              },
            },
          ],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, $tools", () => {
    const rendered = render({
      prompt: "mock-prompt",
      $tools: [
        {
          name: "mock-function-1",
          description: "mock-description-1",
          parameters: [
            {
              name: "mock-function-1-param-1",
              description: "mock-function-1-param-1-description-1",
              type: "STR",
              required: false,
            },
            {
              name: "mock-function-1-param-2",
              description: "mock-function-1-param-2-description-2",
              type: "NUM",
              required: false,
            },
          ],
          invocations: [],
        },
        {
          name: "mock-function-2",
          description: "mock-description-2",
          parameters: [
            {
              name: "mock-function-2-param-1",
              description: "mock-function-2-param-1-description-1",
              type: "BOOL",
              required: false,
            },
          ],
          invocations: [],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, tools, $tools", () => {
    const rendered = render({
      prompt: "mock-prompt",
      tools: [
        {
          function_declarations: [
            {
              name: "mock-function",
              description: "mock-description",
              parameters: {
                type: "OBJECT",
                properties: {
                  key: { type: "STRING" },
                },
              },
            },
          ],
        },
      ],
      $tools: [
        {
          name: "mock-function-1",
          description: "mock-description-1",
          parameters: [
            {
              name: "mock-function-1-param-1",
              description: "mock-function-1-param-1-description-1",
              type: "STR",
              required: false,
            },
            {
              name: "mock-function-1-param-2",
              description: "mock-function-1-param-2-description-2",
              type: "NUM",
              required: true,
            },
          ],
          invocations: [],
        },
        {
          name: "mock-function-2",
          description: "mock-description-2",
          parameters: [
            {
              name: "mock-function-2-param-1",
              description: "mock-function-2-param-1-description-1",
              type: "BOOL",
              required: false,
            },
          ],
          invocations: [],
        },
      ],
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, tools_config", () => {
    const rendered = render({
      prompt: "mock-prompt",
      tools_config: {
        mode: "AUTO",
        allowed_function_names: ["mock-function"],
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  /**
   * "Native" options:
   */
  test("prompt, safety_settings", () => {
    const rendered = render({
      prompt: "mock-prompt",
      safety_settings: {
        category: "mock-category",
        threshold: "mock-threshold",
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, generation_config", () => {
    const rendered = render({
      prompt: "mock-prompt",
      generation_config: {
        temperature: 0.7,
        top_p: 0.9,
      },
    });
    expect(rendered).toMatchSnapshot();
  });

  test("prompt, tool, tools_config, system_instruction, safety_settings, generation_config", () => {
    const rendered = render({
      prompt: "mock-prompt",
      contents: [
        {
          role: "model",
          parts: [{ text: "mock-model-text" }],
        },
      ],
      tools: [
        {
          function_declarations: [
            {
              name: "mock-function",
            },
          ],
        },
      ],
      tools_config: {
        mode: "ANY",
      },
      system_instruction: {
        parts: [{ text: "mock-system-text" }],
      },
      safety_settings: {
        threshold: "mock-threshold",
        method: "mock-method",
      },
      generation_config: {
        top_p: 0.7,
        max_output_tokens: 100,
      },
    });
    expect(rendered).toMatchSnapshot();
  });
});
