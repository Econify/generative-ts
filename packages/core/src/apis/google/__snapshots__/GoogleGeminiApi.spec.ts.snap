// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleGeminiApi.requestTemplate $prompt 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, $tools 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description-1",
          "name": "mock-function-1",
          "parameters": {
            "properties": {
              "mock-function-1-param-1": {
                "description": "mock-function-1-param-1-description-1",
                "type": "STRING",
              },
              "mock-function-1-param-2": {
                "description": "mock-function-1-param-2-description-2",
                "type": "NUMBER",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
        {
          "description": "mock-description-2",
          "name": "mock-function-2",
          "parameters": {
            "properties": {
              "mock-function-2-param-1": {
                "description": "mock-function-2-param-1-description-1",
                "type": "BOOLEAN",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents ending with function_call, $tools with matching invocation (appends function_response content items) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "functionCall": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "function_response": {
            "name": "mock-function",
            "response": {
              "returned": {
                "responseKey": "responseValue",
              },
            },
          },
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description",
          "name": "mock-function",
          "parameters": {
            "properties": {
              "key": {
                "description": "mock-key-description",
                "type": "STRING",
              },
            },
            "required": [
              "key",
            ],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents ending with model function_call, $tools without matching invocation (appends $prompt; TODO logs warning) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "functionCall": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "another-description",
          "name": "another-function",
          "parameters": {
            "properties": {
              "another-key": {
                "description": "another-key-description",
                "type": "STRING",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents ending with model function_call, no $tools (prepends and appends $prompt; TODO logs warning) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "functionCall": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents ending with user function_response (prepends $prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "functionCall": {
            "args": {
              "key": "value",
            },
            "name": "mock-function",
          },
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "function_response": {
            "name": "mock-function",
            "response": {
              "responseKey": "responseValue",
            },
          },
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents starting and ending with model (appends AND prepends $prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text-2",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents starting and ending with user (does not add $prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text-2",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents with model / user (prepends $prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text-2",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents with user / model (appends $prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, contents, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, examplePairs 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-msg-2",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-2",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, examplePairs, contents with model / user (inserts $prompt, conversation is valid) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-example-pair",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-example-pair",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, examplePairs, contents with user / model (appends $prompt) 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-example-pair",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-example-pair",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-user-text",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, examplePairs, contents, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, examplePairs, system 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, examplePairs, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-user-msg-1",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-assistant-msg-1",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, generation_config 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "generation_config": {
    "temperature": 0.7,
    "top_p": 0.9,
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, safety_settings 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "safety_settings": {
    "category": "mock-category",
    "threshold": "mock-threshold",
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, system 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, system, contents 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, system, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
      {
        "text": "mock-additional-instruction",
      },
      {
        "text": "mock-additional-instruction-2",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, system_instruction 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, tool, tools_config, system_instruction, safety_settings, generation_config 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
    {
      "parts": [
        {
          "text": "mock-model-text",
        },
      ],
      "role": "model",
    },
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "generation_config": {
    "candidate_count": 5,
    "frequency_penalty": 0.3,
    "max_output_tokens": 100,
    "presence_penalty": 0.5,
    "response_mime_type": "application/json",
    "stop_sequences": [
      "mock-stop-sequence",
    ],
    "temperature": 0.7,
    "top_k": 10,
    "top_p": 0.9,
  },
  "safety_settings": {
    "category": "mock-category",
    "max_influential_terms": 10,
    "method": "mock-method",
    "threshold": "mock-threshold",
  },
  "system_instruction": {
    "parts": [
      {
        "text": "mock-system-text",
      },
    ],
  },
  "tools": [
    {
      "function_declarations": [
        {
          "name": "mock-function",
        },
      ],
    },
  ],
  "tools_config": {
    "allowed_function_names": [
      "mock-function",
    ],
    "mode": "ANY",
  },
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, tools 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description",
          "name": "mock-function",
          "parameters": {
            "properties": {
              "key": {
                "type": "STRING",
              },
            },
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, tools, $tools 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools": [
    {
      "function_declarations": [
        {
          "description": "mock-description",
          "name": "mock-function",
          "parameters": {
            "properties": {
              "key": {
                "type": "STRING",
              },
            },
            "type": "OBJECT",
          },
        },
      ],
    },
    {
      "function_declarations": [
        {
          "description": "mock-description-1",
          "name": "mock-function-1",
          "parameters": {
            "properties": {
              "mock-function-1-param-1": {
                "description": "mock-function-1-param-1-description-1",
                "type": "STRING",
              },
              "mock-function-1-param-2": {
                "description": "mock-function-1-param-2-description-2",
                "type": "NUMBER",
              },
            },
            "required": [
              "mock-function-1-param-2",
            ],
            "type": "OBJECT",
          },
        },
        {
          "description": "mock-description-2",
          "name": "mock-function-2",
          "parameters": {
            "properties": {
              "mock-function-2-param-1": {
                "description": "mock-function-2-param-1-description-1",
                "type": "BOOLEAN",
              },
            },
            "required": [],
            "type": "OBJECT",
          },
        },
      ],
    },
  ],
}
`;

exports[`GoogleGeminiApi.requestTemplate $prompt, tools_config 1`] = `
{
  "contents": [
    {
      "parts": [
        {
          "text": "mock-prompt",
        },
      ],
      "role": "user",
    },
  ],
  "tools_config": {
    "allowed_function_names": [
      "mock-function",
    ],
    "mode": "AUTO",
  },
}
`;
